File: ProductExtractor/instructions.py
from models.category import Category
from abc import ABC, abstractmethod
from typing import Callable

from selenium.common import TimeoutException
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.common.window import WindowTypes
from models.provider import Provider

from models.product import ProductBuilder
from models.offer import Offer, OfferBuilder

from browser import Browser

DEFAULT_TIMEOUT = 10


class BrowserInstruction[T](ABC):
    @abstractmethod
    def accept(self, browser: Browser) -> T:
        pass


class GotoURL(BrowserInstruction[None]):
    def __init__(self, url: str):
        self.url = url

    def accept(self, browser: Browser) -> None:
        browser.driver.get(self.url)


class OpenTab(BrowserInstruction[str]):
    def accept(self, browser: Browser) -> str:
        original = browser.driver.current_window_handle
        browser.driver.switch_to.new_window(WindowTypes.TAB)
        return original


class CloseTab(BrowserInstruction[None]):
    def __init__(self, previous: str) -> None:
        self.prev = previous

    def accept(self, browser: Browser) -> None:
        browser.driver.close()
        browser.driver.switch_to.window(self.prev)


type Processing[T] = Callable[[str], T]


def identity(T):
    return T


class Scrape[T = str](BrowserInstruction[list[T]]):
    def __init__(
        self,
        selector: str,
        attribute_name: str,
        processing: Processing[T] = identity,
    ):
        self.selector = selector
        self.attribute_name = attribute_name
        self.processing = processing

    def accept(self, browser: Browser) -> list[T]:
        try:
            elements = WebDriverWait(browser.driver, DEFAULT_TIMEOUT).until(
                EC.presence_of_all_elements_located((By.CSS_SELECTOR, self.selector))
            )
        except TimeoutException:
            print("selector didnt match:", self.selector)
            return []

        data: list[str] = []

        for element in elements:
            attr = element.get_attribute(self.attribute_name)
            if attr is None:
                continue
            data.append(attr)

        return list(map(self.processing, data))


class ScrapeProductInfo(BrowserInstruction):
    def __init__(
        self,
        scrape_keys: Scrape[str],
        scrape_values: Scrape[str],
    ) -> None:
        self.scrape_keys = scrape_keys
        self.scrape_values = scrape_values

    def accept(self, browser: Browser) -> dict[str, str]:
        keys = self.scrape_keys.accept(browser)
        values = self.scrape_values.accept(browser)

        return dict(zip(keys, values))


class ScrapeOffers(Scrape[OfferBuilder]):
    def __init__(
        self,
        scrape_prices: Scrape[float],
        scrape_names: Scrape[str],
        scrape_links: Scrape[str],
        scrape_images: Scrape[str],
        scrape_references: Scrape[str],
        scrape_product_info: ScrapeProductInfo,
    ) -> None:
        self.scrape_prices = scrape_prices
        self.scrape_names = scrape_names
        self.scrape_links = scrape_links
        self.scrape_images = scrape_images
        self.scrape_references = scrape_references
        self.scrape_product_info = scrape_product_info

    def accept(self, browser: Browser) -> list[OfferBuilder]:
        prices = browser.execute(self.scrape_prices)
        names = browser.execute(self.scrape_names)
        links = browser.execute(self.scrape_links)
        images = browser.execute(self.scrape_images)
        references = browser.execute(self.scrape_references)

        offers = []
        for price, name, link, image, ref in zip(
            prices, names, links, images, references
        ):
            product_info = browser.execute(
                GetProductInfo(link, self.scrape_product_info)
            )
            product = ProductBuilder(ref, name)
            product.set_image(image)
            product.set_info(product_info)

            offer = OfferBuilder(product, price, link)

            offers.append(offer)

        return offers


class ScrapeCategory(Scrape[OfferBuilder]):
    def __init__(self, category: Category, scrape_offers: ScrapeOffers) -> None:
        self.category = category
        self.scrape_offers = scrape_offers

    def accept(self, browser: Browser) -> list[OfferBuilder]:
        offers = browser.execute(self.scrape_offers)
        for offer in offers:
            offer.product.set_category(self.category)

        return offers


class GetCategory(BrowserInstruction[list[OfferBuilder]]):
    def __init__(self, url: str, scrape_category: ScrapeCategory) -> None:
        self.url = url
        self.scrape_category = scrape_category

    def accept(self, browser: Browser) -> list[OfferBuilder]:
        browser.execute(GotoURL(self.url))
        offers = browser.execute(self.scrape_category)
        return offers


class ScrapeProvider(Scrape[Offer]):
    def __init__(self, provider: Provider, get_category: GetCategory) -> None:
        self.provider = provider
        self.get_category = get_category

    def accept(self, browser: Browser) -> list[Offer]:
        offers = browser.execute(self.get_category)

        for offer in offers:
            offer.set_provider(self.provider)

        return [offer.build() for offer in offers]


class GetOffers(BrowserInstruction):
    def __init__(self, url: str, scrape_instruction: ScrapeOffers) -> None:
        self.url = url
        self.scrape_instruction = scrape_instruction

    def accept(self, browser: Browser) -> list[OfferBuilder]:
        browser.execute(GotoURL(self.url))
        return browser.execute(self.scrape_instruction)


class GetProductInfo(BrowserInstruction):
    def __init__(
        self, product_url: str, scrape_instructions: ScrapeProductInfo
    ) -> None:
        self.url = product_url
        self.scrape_instruction = scrape_instructions

    def accept(self, browser: Browser) -> dict[str, str]:
        origibal = browser.execute(OpenTab())
        browser.execute(GotoURL(self.url))
        info = browser.execute(self.scrape_instruction)
        browser.execute(CloseTab(origibal))
        return info

File: offer-extractor/instructions.ts
import { By, until, WebElement } from "selenium-webdriver";
import { Browser } from "./browser.ts";
import { Offer, OfferBuilder } from "./models/offer.ts";
import { ProductBuilder } from "./models/product.ts";
import { Category } from "./models/category.ts";
import { Provider } from "./models/provider.ts";

const DEFAULT_TIMEOUT = 10000;

export interface Instruction<T> {
    accept(browser: Browser): Promise<T>;
}


export class GotoURL implements Instruction<void> {
    url: string;

    constructor(url: string) {
        this.url = url;
    }

    async accept(browser: Browser): Promise<void> {
        await browser.driver.get(this.url);
    }
}

export class OpenTab implements Instruction<string> {
    async accept(browser: Browser): Promise<string> {
        const original = await browser.driver.getWindowHandle();
        await browser.driver.switchTo().newWindow('tab');
        return original;
    }
}

export class CloseTab implements Instruction<void> {
    previous: string;

    constructor(previous: string) {
        this.previous = previous;
    }

    async accept(browser: Browser): Promise<void> {
        await browser.driver.close()
        await browser.driver.switchTo().window(this.previous)
    }
}

type Processing<T> = (str: string) => T;

export function identity(str: string): string {
    return str;
}


export class Scrape<T = string> implements Instruction<T[]> {
    selector: string;
    attribute: string;
    processing: Processing<T>;

    constructor(
        selector: string,
        attribute: string,
        processing: Processing<T>,
    ) {
        this.selector = selector;
        this.attribute = attribute;
        this.processing = processing;
    }


    async accept(browser: Browser): Promise<T[]> {
        let elements: WebElement[];
        try {
            elements = await browser.driver.wait(
                until.elementsLocated(By.css(this.selector)),
                DEFAULT_TIMEOUT,
            );
        } catch (e) {
            console.log("selector didnt match:", this.selector)
            elements = [];
        }


        let data = elements
            .map(async (element) => await element.getAttribute(this.attribute))
            .filter(data => data != null)
            .map(async datum => this.processing(await datum));

        return Promise.all(data);
    }
}


export class ScrapeProductInfo implements Instruction<Map<string, string>> {
    scrapeKeys: Scrape<string>;
    scrapeValues: Scrape<string>;

    constructor(scrapeKeys: Scrape<string>, scrapeValues: Scrape<string>) {
        this.scrapeKeys = scrapeKeys;
        this.scrapeValues = scrapeValues;
    }

    async accept(browser: Browser): Promise<Map<string, string>> {
        const keys = await this.scrapeKeys.accept(browser);
        const values = await this.scrapeValues.accept(browser);
        const productInfo = new Map();

        keys.forEach((key, i) => productInfo.set(key, values[i]));
        return productInfo;
    }
}


export class ScrapeOffers {
    scrapePrices: Scrape<number>;
    scrapeNames: Scrape<string>;
    scrapeLinks: Scrape<string>;
    scrapeImages: Scrape<string>;
    scrapeReferences: Scrape<string>;
    scrapeProductInfo: ScrapeProductInfo;

    constructor(
        scrapePrices: Scrape<number>,
        scrapeNames: Scrape<string>,
        scrapeLinks: Scrape<string>,
        scrapeImages: Scrape<string>,
        scrapeReferences: Scrape<string>,
        scrapeProductInfo: ScrapeProductInfo,
    ) {
        this.scrapePrices = scrapePrices;
        this.scrapeNames = scrapeNames;
        this.scrapeLinks = scrapeLinks;
        this.scrapeImages = scrapeImages;
        this.scrapeReferences = scrapeReferences;
        this.scrapeProductInfo = scrapeProductInfo;
    }

    async accept(browser: Browser): Promise<OfferBuilder[]> {
        const prices = await browser.execute(this.scrapePrices)
        const names = await browser.execute(this.scrapeNames)
        const links = await browser.execute(this.scrapeLinks)
        const images = await browser.execute(this.scrapeImages)
        const references = await browser.execute(this.scrapeReferences)

        const offers: OfferBuilder[] = [];
        for (let i = 0; i < names.length; i++) {
            const productInfo = await browser.execute(
                new GetProductInfo(links[i]!, this.scrapeProductInfo)
            );

            const product = new ProductBuilder(references[i]!, names[i]!);
            product.setImage(images[i]!);
            product.setInfo(productInfo);

            const offer = new OfferBuilder(product, prices[i]!, links[i]!)

            offers.push(offer);

        }

        return offers
    }
}

export class ScrapeCategory implements Instruction<OfferBuilder[]> {
    category: Category;
    scrapeOffers: ScrapeOffers;

    constructor(category: Category, scrapeOffers: ScrapeOffers) {
        this.category = category
        this.scrapeOffers = scrapeOffers
    }

    async accept(browser: Browser): Promise<OfferBuilder[]> {
        const offers = await browser.execute(this.scrapeOffers);
        offers.forEach(offer => offer.product.setCategory(this.category));

        return offers
    }
}


export class GetCategory implements Instruction<OfferBuilder[]> {
    url: string;
    scrapeCategory: ScrapeCategory;


    constructor(url: string, scrapeCategory: ScrapeCategory) {
        this.url = url
        this.scrapeCategory = scrapeCategory
    }

    async accept(browser: Browser): Promise<OfferBuilder[]> {
        await browser.execute(new GotoURL(this.url))
        const offers = await browser.execute(this.scrapeCategory)
        return offers
    }
}


export class ScrapeProvider implements Instruction<Offer[]> {
    provider: Provider;
    getCategories: GetCategory[];


    constructor(provider: Provider, ...getCategories: GetCategory[]) {
        this.provider = provider
        this.getCategories = getCategories
    }

    async accept(browser: Browser): Promise<Offer[]> {
        const offers: Offer[] = [];


        for (const getCategory of this.getCategories) {
            const categoryOffers = await browser.execute(getCategory);
            for (const categoryOffer of categoryOffers) {
                categoryOffer.setProvider(this.provider);
                offers.push(categoryOffer.build());
            }
        }

        return offers;
    }


}

export class GetOffers implements Instruction<OfferBuilder[]> {
    url: string;
    scrapeInstruction: ScrapeOffers;

    constructor(url: string, scrapeInstruction: ScrapeOffers) {
        this.url = url
        this.scrapeInstruction = scrapeInstruction
    }

    async accept(browser: Browser): Promise<OfferBuilder[]> {
        await browser.execute(new GotoURL(this.url))
        return await browser.execute(this.scrapeInstruction)
    }
}

export class GetProductInfo implements Instruction<Map<string, string>> {
    url: string;
    scrapeProductInfo: ScrapeProductInfo;

    constructor(url: string, scrapeProductInfo: ScrapeProductInfo) {
        this.url = url
        this.scrapeProductInfo = scrapeProductInfo
    }

    async accept(browser: Browser): Promise<Map<string, string>> {
        const original = await browser.execute(new OpenTab())
        await browser.execute(new GotoURL(this.url))
        const info = await browser.execute(this.scrapeProductInfo)
        await browser.execute(new CloseTab(original))
        return info;
    }
}
